civis/service_client.py

```diff
@@ -0,0 +1,143 @@
+
+    def __init__(self, session_kwargs, client,
+                 return_type='civis', root_path=None):
+        self._session_kwargs = session_kwargs
+        self._return_type = return_type
+        self._client = client
+        self._root_path = root_path
+
+    def _build_path(self, path):
+        if not path:
+            return self._client._base_url
+        if self._root_path is None:
+            return tostr_urljoin(self._client._base_url, path.strip("/"))
+        return tostr_urljoin(self._client._base_url,
+                             self._root_path.strip("/"),
+                             path.strip("/"))
+
+    def _make_request(self, method, path=None, params=None, data=None,
+                      **kwargs):
+        url = self._build_path(path)
+
+        with requests.Session() as sess:
+            auth_service_session(sess, self._client._service_id)
+            with self._lock:
+                response = sess.request(method, url, json=data,
+                                        params=params, **kwargs)
+
+        if not response.ok:
+            six.raise_from(ValueError(response.text),
+                           ValueError)
+
+        return response
+
+
+class ServiceClient():
+
+    def __init__(self, service_id, root_path=None, swagger_path="/endpoints"):
+        return_type = 'snake'
+        self._session_kwargs = {}
+        self._service_id = service_id
+        self._base_url = self.get_base_url()
+        self._root_path = root_path
+        self._swagger_path = swagger_path
+        # Catch deprecation warnings from generate_classes_maybe_cached and
+        # the functions it calls until the `resources` argument is removed.
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore",
+                category=FutureWarning,
+                module='civis')
+            classes = self.generate_classes()
+        for class_name, cls in classes.items():
+            setattr(self, class_name, cls(self._session_kwargs, client=self,
+                                          return_type=return_type,
+                                          root_path=root_path))
+
+    def parse_path(self, path, operations):
+        """ Parse an endpoint into a class where each valid http request
+        on that endpoint is converted into a convenience function and
+        attached to the class as a method.
+        """
+        if self._root_path is not None:
+            path = path.replace(self._root_path, '')
+        path = path.strip('/')
+        modified_base_path = re.sub("-", "_", path.split('/')[0].lower())
+        methods = []
+        for verb, op in operations.items():
+            method = parse_method(verb, op, path)
+            if method is None:
+                continue
+            methods.append(method)
+        return modified_base_path, methods
+
+    def parse_api_spec(self, api_spec):
+        paths = api_spec['paths']
```

Comment:
this only works for swagger 2.0 so it should have ['paths'], I wanted to add some validation but was feeling unsure about what to validate. I think I'm going to use this library to validate instead of a custom function https://github.com/p1c2u/openapi-spec-validator

Reply:
what I'm wondering is if a service changed their documentation style, how much would break inside the the python client? Would that service no longer be callable (which is kind of ok) or would a KeyError cause all of civis-python to be unusable?

Reply:
It all relies on consistent documentation style. If they change the style then the client can't parse out the methods and such. The client uses the spec to create the classes and methods.

The point of swagger/openAPI was to make a consistent api document format. But it was kind of a headache researching libraries because they seem to only support version 2.0 *or* 3.0. Civis API is using version 2.0, and flask-swagger that we're using on survey-toolbox is using version 2.0. If we were to mess up the syntax on survey_toolbox it would also fail to generate the documentation and show an error on the page.

Reply:
@leeacto  I wanted your opinion, I added validation [here](https://github.com/ggarcia-civis/civis-python/blob/e042562f17e2ea17cfc364a29b867ce68edf1a3c/civis/service_client.py#L129), but it will require adding the package to the requirements.txt file. Do you think it's worth another install to have the validation on something that's already standardized?

Reply:
talked offline:

v2 validator shouldn't be necessary. Any issues that arise will break in only that Service anyway, allowing the rest of civis-python to be usable.